File: Gateway/byes/plan_executor.py
from __future__ import annotations

from typing import Any, Callable


UiEmitFn = Callable[[dict[str, Any]], None]
NowMsFn = Callable[[], int]


def execute_plan(
    plan: dict[str, Any],
    emit_event_fn: UiEmitFn,
    now_ms_fn: NowMsFn,
) -> dict[str, Any]:
    actions_raw = plan.get("actions")
    actions_raw = actions_raw if isinstance(actions_raw, list) else []
    indexed_actions = [
        (idx, dict(action))
        for idx, action in enumerate(actions_raw)
        if isinstance(action, dict)
    ]
    ordered = sorted(
        indexed_actions,
        key=lambda pair: (_as_int(pair[1].get("priority"), 9999), pair[0]),
    )

    executed: list[dict[str, Any]] = []
    blocked: list[dict[str, Any]] = []
    pending_confirms: list[dict[str, Any]] = []
    ui_commands: list[dict[str, Any]] = []

    blocking_priority: int | None = None
    for default_index, action in ordered:
        action_type = str(action.get("type", "")).strip().lower()
        priority = _as_int(action.get("priority"), 9999)
        action_id = _resolve_action_id(action, default_index + 1)
        payload = action.get("payload")
        payload = payload if isinstance(payload, dict) else {}

        if _is_blocked_by_previous(blocking_priority, priority, action_type):
            blocked_item = {
                "type": action_type or "unknown",
                "actionId": action_id,
                "reason": "blocked_by_previous_blocking_action",
            }
            blocked.append(blocked_item)
            continue

        if action_type == "confirm":
            confirm_id = str(payload.get("confirmId", "")).strip() or f"confirm-{action_id}"
            timeout_ms = _as_int(payload.get("timeoutMs"), 5000)
            text = str(payload.get("text", "")).strip() or "Please confirm."
            command = {
                "kind": "ui.confirm_request",
                "actionId": action_id,
                "confirmId": confirm_id,
                "text": text,
                "timeoutMs": timeout_ms,
                "tsMs": int(max(0, now_ms_fn())),
            }
            ui_commands.append(command)
            pending_confirms.append(
                {
                    "confirmId": confirm_id,
                    "timeoutMs": timeout_ms,
                    "actionId": action_id,
                }
            )
            executed.append({"type": action_type, "actionId": action_id})
            emit_event_fn(command)
        elif action_type in {"speak", "overlay", "haptic", "stop"}:
            command = {
                "kind": "ui.command",
                "commandType": action_type,
                "actionId": action_id,
                "text": str(payload.get("text", "")).strip() if payload.get("text") is not None else "",
                "label": str(payload.get("label", "")).strip() if payload.get("label") is not None else "",
                "reason": str(payload.get("reason", "")).strip() if payload.get("reason") is not None else "",
                "tsMs": int(max(0, now_ms_fn())),
            }
            ui_commands.append(command)
            executed.append({"type": action_type, "actionId": action_id})
            emit_event_fn(command)
        else:
            blocked.append(
                {
                    "type": action_type or "unknown",
                    "actionId": action_id,
                    "reason": "unsupported_action_type",
                }
            )
            continue

        if bool(action.get("blocking")):
            blocking_priority = priority

    return {
        "ok": True,
        "executed": executed,
        "blocked": blocked,
        "pendingConfirms": pending_confirms,
        "uiCommands": ui_commands,
        "executedCount": len(executed),
        "blockedCount": len(blocked),
        "pendingConfirmCount": len(pending_confirms),
    }


def _is_blocked_by_previous(blocking_priority: int | None, current_priority: int, action_type: str) -> bool:
    if blocking_priority is None:
        return False
    if current_priority > blocking_priority:
        return True
    if current_priority == blocking_priority and action_type not in {"confirm", "stop"}:
        return True
    return False


def _resolve_action_id(action: dict[str, Any], ordinal: int) -> str:
    action_id = str(action.get("actionId", "")).strip()
    if action_id:
        return action_id
    payload = action.get("payload")
    if isinstance(payload, dict):
        payload_id = str(payload.get("actionId", "")).strip()
        if payload_id:
            return payload_id
    return f"a{int(max(1, ordinal))}"


def _as_int(value: Any, default: int) -> int:
    try:
        if value is None or isinstance(value, bool):
            return int(default)
        return int(float(value))
    except Exception:
        return int(default)

File: Assets/Scripts/BYES/Plan/PlanExecutor.cs
using System;
using BYES.Core;
using BYES.Telemetry;
using BYES.UI;
using UnityEngine;

namespace BYES.Plan
{
    public class PlanExecutor : MonoBehaviour
    {
        [Serializable]
        public class ActionRef
        {
            public string type;
            public string actionId;
            public string reason;
        }

        [Serializable]
        public class PendingConfirm
        {
            public string confirmId;
            public int timeoutMs;
            public string actionId;
        }

        [Serializable]
        public class UiCommand
        {
            public string kind;
            public string commandType;
            public string actionId;
            public string text;
            public string label;
            public string reason;
            public string confirmId;
            public int timeoutMs;
        }

        [Serializable]
        public class ExecutionSummary
        {
            public bool ok;
            public int executedCount;
            public int blockedCount;
            public int pendingConfirmCount;
            public ActionRef[] executed;
            public ActionRef[] blocked;
            public PendingConfirm[] pendingConfirms;
            public UiCommand[] uiCommands;
        }

        public bool IsStopped { get; private set; }
        public string LastOverlayText { get; private set; } = string.Empty;

        private string _runIdForAck = "unknown-run";
        private int _frameSeqForAck = 1;

        public void SetExecutionContext(string runId, int frameSeq)
        {
            _runIdForAck = string.IsNullOrWhiteSpace(runId) ? "unknown-run" : runId.Trim();
            _frameSeqForAck = Mathf.Max(1, frameSeq);
        }

        public void ExecuteSummary(ExecutionSummary summary, Action<string, bool> onConfirmDecision)
        {
            if (summary == null)
            {
                Debug.LogWarning("[PlanExecutor] execution summary is null");
                return;
            }

            if (summary.uiCommands != null && summary.uiCommands.Length > 0)
            {
                foreach (var cmd in summary.uiCommands)
                {
                    if (cmd == null)
                    {
                        continue;
                    }
                    ExecuteCommand(cmd, onConfirmDecision);
                }
            }

            if (summary.pendingConfirms != null && summary.pendingConfirms.Length > 0)
            {
                var pending = summary.pendingConfirms[0];
                if (pending != null)
                {
                    ShowConfirmFromPending(pending, onConfirmDecision);
                }
            }
        }

        private void ExecuteCommand(UiCommand command, Action<string, bool> onConfirmDecision)
        {
            var kind = (command.kind ?? string.Empty).Trim().ToLowerInvariant();
            if (kind == "ui.confirm_request")
            {
                var prompt = string.IsNullOrWhiteSpace(command.text) ? "Please confirm." : command.text;
                var timeoutMs = command.timeoutMs > 0 ? command.timeoutMs : 5000;
                ShowConfirm(command.confirmId, prompt, timeoutMs, onConfirmDecision);
                return;
            }

            var commandType = (command.commandType ?? string.Empty).Trim().ToLowerInvariant();
            switch (commandType)
            {
                case "speak":
                    Debug.Log($"[PlanExecutor] SPEAK: {command.text}");
                    ByesFrameTelemetry.AckFeedback(
                        _runIdForAck,
                        _frameSeqForAck,
                        "tts",
                        true,
                        ByesFrameTelemetry.NowUnixMs(),
                        providerBackend: "android_tts",
                        providerModel: "quest-tts",
                        providerDevice: "quest",
                        providerReason: "client_tts",
                        providerIsMock: false);
                    break;
                case "overlay":
                case "ar":
                    LastOverlayText = command.label ?? string.Empty;
                    ByesOverlayRenderer.EnsureExists().RenderOverlayCommand(
                        _runIdForAck,
                        _frameSeqForAck,
                        commandType,
                        command.label,
                        command.text,
                        command.reason
                    );
                    break;
                case "haptic":
                    if (ByesHaptics.Instance.TrySendPulse(HapticChannel.Both, 0.5f, 0.08f, command.actionId, command.confirmId))
                    {
                        if (ByesOverlayAckThrottler.Instance.TryMark(_runIdForAck, _frameSeqForAck, "haptic"))
                        {
                            ByesFrameTelemetry.AckFeedback(_runIdForAck, _frameSeqForAck, "haptic", true, ByesFrameTelemetry.NowUnixMs());
                        }
                    }
                    else if (Debug.isDebugBuild)
                    {
                        Debug.Log("[PlanExecutor] HAPTIC skipped (unsupported or debounced)");
                    }
                    break;
                case "stop":
                    IsStopped = true;
                    ByesOverlayRenderer.EnsureExists().RenderStop(_runIdForAck, _frameSeqForAck, string.IsNullOrWhiteSpace(command.text) ? "STOP" : command.text);
                    if (ByesHaptics.Instance.TrySendPulse(HapticChannel.Both, 0.9f, 0.15f, command.actionId, command.confirmId))
                    {
                        if (ByesOverlayAckThrottler.Instance.TryMark(_runIdForAck, _frameSeqForAck, "haptic"))
                        {
                            ByesFrameTelemetry.AckFeedback(_runIdForAck, _frameSeqForAck, "haptic", true, ByesFrameTelemetry.NowUnixMs());
                        }
                    }
                    Debug.Log($"[PlanExecutor] STOP: {command.reason}");
                    break;
                default:
                    Debug.LogWarning($"[PlanExecutor] unknown ui command type={commandType}");
                    break;
            }
        }

        private void ShowConfirmFromPending(PendingConfirm pending, Action<string, bool> onConfirmDecision)
        {
            var confirmId = pending != null ? pending.confirmId : string.Empty;
            var timeoutMs = pending != null ? pending.timeoutMs : 5000;
            var actionId = pending != null ? pending.actionId : string.Empty;
            var prompt = string.IsNullOrWhiteSpace(actionId) ? "Please confirm." : actionId;
            ShowConfirm(confirmId, prompt, timeoutMs, onConfirmDecision);
        }

        private void ShowConfirm(string confirmId, string prompt, int timeoutMs, Action<string, bool> onConfirmDecision)
        {
            var safeConfirmId = string.IsNullOrWhiteSpace(confirmId) ? $"confirm-{Time.frameCount}" : confirmId.Trim();
            var safePrompt = string.IsNullOrWhiteSpace(prompt) ? "Please confirm." : prompt.Trim();
            var safeTimeout = timeoutMs > 0 ? timeoutMs : 5000;

            var state = ByesSystemState.Instance;
            if (state != null)
            {
                state.SetPendingConfirm(1, safeConfirmId);
            }

            ByesConfirmPanel.EnsureExists().ShowConfirm(
                _runIdForAck,
                _frameSeqForAck,
                safeConfirmId,
                safePrompt,
                safeTimeout,
                onDecision: (id, accepted) =>
                {
                    if (state != null)
                    {
                        state.SetPendingConfirm(0, id);
                    }
                    onConfirmDecision?.Invoke(id, accepted);
                }
            );
        }
    }
}
